lib/manageiq/appliance_console/auth_utilities.rb

```diff
@@ -0,0 +1,87 @@
+require "uri"
+require "erb"
+
+module ManageIQ
+  module ApplianceConsole
+    module AuthUtilities
+      HTTPD_CONFIG_DIRECTORY = Pathname.new("/etc/httpd/conf.d")
+
+      def restart_httpd
+        httpd_service = LinuxAdmin::Service.new("httpd")
+        if httpd_service.running?
+          say("Restarting httpd ...")
+          httpd_service.restart
+        end
+      end
+
+      def path_is_file?(path)
+        path.present? && !path_is_url?(path)
+      end
+
+      def path_is_url?(path)
+        path =~ /\A#{URI.regexp(["http", "https"])}\z/x
+      end
+
+      # File Management
+
+      def remove_file(path)
+        if path.exist?
+          debug_msg("Removing #{path} ...")
+          File.delete(path)
+        end
+      end
+
+      def copy_template(dir, file, template_parameters = nil)
+        src_path = template_directory.join(relative_from_root(dir), file)
+        dest_path = dir.join(file)
+        dest_path = dest_path.sub_ext('') if src_path.extname == ".erb"
+        debug_msg("Copying template #{src_path} to #{dest_path} ...")
+        if src_path.extname == ".erb"
+          template = ERB.new(File.read(src_path), nil, '-')
+          if template_parameters
+            File.write(dest_path, template.result_with_hash(template_parameters))
+          else
+            File.write(dest_path, template.result(binding))
+          end
+        else
+          FileUtils.cp(src_path, dest_path)
+        end
+      end
+
+      def template_directory
+        @template_directory ||= Pathname.new(ENV.fetch("APPLIANCE_TEMPLATE_DIRECTORY"))
+      end
+
+      def relative_from_root(path)
+        path.absolute? ? path.relative_path_from(Pathname.new("/")) : path
+      end
+
+      # Appliance Settings
+
+      def configure_auth_settings_database
```

Comment:
Perhaps method `configure_auth_settings_database` could also be used to set openidc and saml, drying up 
https://github.com/ManageIQ/manageiq-appliance_console/blob/master/lib/manageiq/appliance_console/saml_authentication.rb#L169
and
https://github.com/ManageIQ/manageiq-appliance_console/pull/104/files#diff-2778f2fd46589a041acba2350d0b78b1R81

into a single method.

I'd ask you consider renaming `configure_auth_settings_database` to `configure_auth_settings` and
 pass arguments to `configure_auth_settings`, which could dry up `configure_auth_setting_database`,  `configure_auth_setting_openidc` and `configure_auth_settings_saml` into a single method.

e.g.:
```ruby
def configure_auth_settings(args)
  puts("Setting Appliance Authentication Settings to #{args['mode']} ...")
  params = [
    "/authentication/mode=#{args['mode']}",
    "/authentication/httpd_role=#{args['role']}",
    "/authentication/saml_enabled=#{args['saml']}",
    "/authentication/oidc_enabled=#{args['oidc']}",
    "/authentication/sso_enabled=#{args['sso']}",
    "/authentication/provider_type=#{args['type']}"
  ]     

  Utilities.rake_run("evm:settings:set", params)
end   


configure_auth_settings('mode' => 'database',
                        'role' => false,
                        'saml' => false,
                        'oidc' => false,
                        'sso'  => false,
                        'type' => 'none' )

configure_auth_settings('mode' => 'httpd',
                        'role' => true,
                        'saml' => true,
                        'oidc' => false,
                        'sso'  => options[:saml_enable_sso] ? 'true' : 'false',
                        'type' => 'saml' )

configure_auth_settings('mode' => 'httpd',
                        'role' => true,
                        'saml' => false,
                        'oidc' => true,
                        'sso'  => options[:saml_enable_sso] ? 'true' : 'false',
                        'type' => 'oidc' )
```

Reply:
well, I would not get rid of configure_auth_settings_database, because then we're duplicating the logic (what params to set for DB mode) in both SAML and OIDC.  We could keep it, but have it call the generic configure_auth_settings. If we go with a generic configure_auth_settings, we would only build params with what is passed in, just in case another type in the future wants to set a different /authentication/param that's not in saml/oidc/....  The only other thing is losing the friendly message "Setting Appliance Authentication Settings to OpenID-Connect ..."  which is replaced above with httpd, unless we send another param to the generic method for the friendly name, or keep the current ones for the say message, which then call the generic configure_auth_settings for doing the work.

Reply:
Maybe I'm just missing something but I don’t understand why you don't want to  get rid of configure_auth_settings_database.

My code segment show all three `database`, `oidc` and `saml` can all use 1 generic method. Am I missing something?

I’m not excited about `puts` in this library because the library could be used by the appliance_console and not only the ap_cli. I would ask that you  limit any invocations to `puts` into the cli specific code.

Reply:
what puts ?

Reply:
Sorry, not puts `say`.  My bad. My point was that I'd rather have a generic utility file not write to stdout. Log to a file perhaps but writing to stdout from a shared utility library doesn't seem like a good idea to me. In this case it could be OK but my philosophy is to avoid writing to stdout from a shared utility.

When setting the authentication configuration, for later use to attempt to diagnose any failures, I'd rather see any message reflect what is actually being set. It's more accurate to log `Setting Appliance Authentication Settings to http` than `Setting Appliance Authentication Settings to OpenID-Connect`. 

Perhaps a:
```ruby
say ("Setting Appliance Authentication Settings for OpenID-Connect support")
configure_auth_settings('mode' => 'httpd',
                        'role' => true,
                        'saml' => false,
                        'oidc' => true,
                        'sso'  => options[:saml_enable_sso] ? 'true' : 'false',
                        'type' => 'oidc' )
```

Reply:
👍 yep, that would work, but with full parameter names, mode, httpd_role, saml_enabled, etc.  for easier extensibility.

Reply:
@jvlcek 58a21749 pushed as per ^, marked as wip, will do a quick test and unwip if all good.

Reply:
Ok, retested with both OpenID-Connect and SAML, still good. so un-wipping.